File: clients/client-python/gravitino/dto/stats/statistic_value_dto.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

"""Data Transfer Objects for statistic values with JSON serialization support."""

from abc import ABC, abstractmethod
from typing import List, Dict, TypeVar

from gravitino.api.statistics.statistic_value import StatisticValue


T = TypeVar("T")


class StatisticValueDTO(StatisticValue[T], ABC):
    """Abstract base class for statistic value DTOs."""

    @abstractmethod
    def to_json(self) -> dict:
        """Convert to JSON representation."""
        pass

    @classmethod
    def from_json(cls, json_dict: dict) -> "StatisticValueDTO":
        """Create a StatisticValueDTO from JSON representation."""
        data_type = json_dict.get("type")
        value = json_dict.get("value")

        if data_type == StatisticValue.Type.BOOLEAN:
            return BooleanValueDTO(value)
        if data_type == StatisticValue.Type.LONG:
            return LongValueDTO(value)
        if data_type == StatisticValue.Type.DOUBLE:
            return DoubleValueDTO(value)
        if data_type == StatisticValue.Type.STRING:
            return StringValueDTO(value)
        if data_type == StatisticValue.Type.LIST:
            items = [StatisticValueDTO.from_json(item) for item in value]
            return ListValueDTO(items)
        if data_type == StatisticValue.Type.OBJECT:
            obj = {k: StatisticValueDTO.from_json(v) for k, v in value.items()}
            return ObjectValueDTO(obj)
        raise ValueError(f"Unknown statistic value type: {data_type}")


class BooleanValueDTO(StatisticValueDTO[bool]):
    """Boolean statistic value DTO."""

    def __init__(self, value: bool):
        self._value = value

    def value(self) -> bool:
        return self._value

    def data_type(self) -> str:
        return StatisticValue.Type.BOOLEAN

    def to_json(self) -> dict:
        return {"type": self.data_type(), "value": self._value}


class LongValueDTO(StatisticValueDTO[int]):
    """Long statistic value DTO."""

    def __init__(self, value: int):
        self._value = value

    def value(self) -> int:
        return self._value

    def data_type(self) -> str:
        return StatisticValue.Type.LONG

    def to_json(self) -> dict:
        return {"type": self.data_type(), "value": self._value}


class DoubleValueDTO(StatisticValueDTO[float]):
    """Double statistic value DTO."""

    def __init__(self, value: float):
        self._value = value

    def value(self) -> float:
        return self._value

    def data_type(self) -> str:
        return StatisticValue.Type.DOUBLE

    def to_json(self) -> dict:
        return {"type": self.data_type(), "value": self._value}


class StringValueDTO(StatisticValueDTO[str]):
    """String statistic value DTO."""

    def __init__(self, value: str):
        self._value = value

    def value(self) -> str:
        return self._value

    def data_type(self) -> str:
        return StatisticValue.Type.STRING

    def to_json(self) -> dict:
        return {"type": self.data_type(), "value": self._value}


class ListValueDTO(StatisticValueDTO[List[StatisticValueDTO]]):
    """List statistic value DTO."""

    def __init__(self, values: List[StatisticValueDTO]):
        self._values = values

    def value(self) -> List[StatisticValueDTO]:
        return self._values

    def data_type(self) -> str:
        return StatisticValue.Type.LIST

    def to_json(self) -> dict:
        return {
            "type": self.data_type(),
            "value": [item.to_json() for item in self._values],
        }


class ObjectValueDTO(StatisticValueDTO[Dict[str, StatisticValueDTO]]):
    """Object statistic value DTO."""

    def __init__(self, values: Dict[str, StatisticValueDTO]):
        self._values = values

    def value(self) -> Dict[str, StatisticValueDTO]:
        return self._values

    def data_type(self) -> str:
        return StatisticValue.Type.OBJECT

    def to_json(self) -> dict:
        return {
            "type": self.data_type(),
            "value": {k: v.to_json() for k, v in self._values.items()},
        }

File: common/src/main/java/org/apache/gravitino/dto/stats/StatisticValueDTO.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.stats;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.google.common.base.Objects;
import java.util.List;
import java.util.Map;
import org.apache.gravitino.rel.types.Type;
import org.apache.gravitino.rel.types.Types;
import org.apache.gravitino.stats.StatisticValue;

/** Base class for Statistic Value Data Transfer Objects. */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
  @JsonSubTypes.Type(value = StatisticValueDTO.BooleanValueDTO.class, name = "boolean"),
  @JsonSubTypes.Type(value = StatisticValueDTO.LongValueDTO.class, name = "long"),
  @JsonSubTypes.Type(value = StatisticValueDTO.DoubleValueDTO.class, name = "double"),
  @JsonSubTypes.Type(value = StatisticValueDTO.StringValueDTO.class, name = "string"),
  @JsonSubTypes.Type(value = StatisticValueDTO.ListValueDTO.class, name = "list"),
  @JsonSubTypes.Type(value = StatisticValueDTO.ObjectValueDTO.class, name = "object")
})
public abstract class StatisticValueDTO<T> implements StatisticValue<T> {

  @JsonProperty("value")
  protected T value;

  protected StatisticValueDTO() {}

  protected StatisticValueDTO(T value) {
    this.value = value;
  }

  @Override
  public T value() {
    return value;
  }

  @Override
  @JsonIgnore
  public abstract Type dataType();

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof StatisticValueDTO)) return false;
    StatisticValueDTO<?> that = (StatisticValueDTO<?>) o;
    return Objects.equal(value, that.value);
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(value);
  }

  /** Boolean value DTO. */
  public static class BooleanValueDTO extends StatisticValueDTO<Boolean> {

    private BooleanValueDTO() {
      super();
    }

    public BooleanValueDTO(Boolean value) {
      super(value);
    }

    @Override
    public Type dataType() {
      return Types.BooleanType.get();
    }
  }

  /** Long value DTO. */
  public static class LongValueDTO extends StatisticValueDTO<Long> {

    private LongValueDTO() {
      super();
    }

    public LongValueDTO(Long value) {
      super(value);
    }

    @Override
    public Type dataType() {
      return Types.LongType.get();
    }
  }

  /** Double value DTO. */
  public static class DoubleValueDTO extends StatisticValueDTO<Double> {

    private DoubleValueDTO() {
      super();
    }

    public DoubleValueDTO(Double value) {
      super(value);
    }

    @Override
    public Type dataType() {
      return Types.DoubleType.get();
    }
  }

  /** String value DTO. */
  public static class StringValueDTO extends StatisticValueDTO<String> {

    private StringValueDTO() {
      super();
    }

    public StringValueDTO(String value) {
      super(value);
    }

    @Override
    public Type dataType() {
      return Types.StringType.get();
    }
  }

  /** List value DTO. */
  public static class ListValueDTO<T> extends StatisticValueDTO<List<StatisticValueDTO<T>>> {

    @JsonProperty("elementType")
    private Type elementType;

    private ListValueDTO() {
      super();
    }

    public ListValueDTO(List<StatisticValueDTO<T>> value, Type elementType) {
      super(value);
      this.elementType = elementType;
    }

    @Override
    public Type dataType() {
      return Types.ListType.of(elementType, false);
    }
  }

  /** Object value DTO. */
  public static class ObjectValueDTO extends StatisticValueDTO<Map<String, StatisticValueDTO<?>>> {

    private ObjectValueDTO() {
      super();
    }

    public ObjectValueDTO(Map<String, StatisticValueDTO<?>> value) {
      super(value);
    }

    @Override
    public Type dataType() {
      // For object values, we use a struct type. The exact fields would be determined by the map
      // keys.
      // For simplicity, we'll return a generic struct type.
      return Types.StructType.of();
    }
  }
}
